add summary field to challenges and guilds

Alys opened the pull request:
This PR:
- Adds a `summary` field to the `challenge` document in the database. The field is not mandatory because old challenges and third-party tools don't have it.
- Uses a pre init hook so that when any challenge without a summary is fetched from the database, the challenge's name is used as the summary. This allows us to assume in the vue code that all challenges will have a summary without us needing to do a migration. The challenge owner can still edit old challenges to add a proper summary of course.
- Adds the summary field to the challenge creation screen.
- Uses a description field in the challenge creation screen for the long description (as in prod).
- Adjusts the API to support an optional summary (we can make it mandatory in a few months after a suitable warning period since it's a breaking change). API will set the summary to be the name if not supplied.
- The mongoose schema enforces a summary character limit of 250. The vue code and API still need to be modified to do the same.


If this is all okay so far, I'll keep going with it and do the same sort of thing for guilds.

Alys committed:
create new summary field for challenges

Alys commented on website/common/locales/en/newClient.json:
```diff
@@ -98,7 +98,6 @@
   "yourNotOnQuest": "You're not on a quest",
   "questDescription": "Quests allow players to focus on long-term, in-game goals with the members of their party.",
   "haveNoChallenges": "You don’t have any Challenges",
-  "challengeDescription": "Challenges are community events in which players compete and earn prizes by completing a group of related tasks.",
```
I've deleted this because `challengeDescription` was used elsewhere for a different string, and that string there is duplicated in `challengeDescription0`.

TheHollidayInn commented on website/server/models/challenge.js:
```diff
@@ -43,6 +44,16 @@ schema.plugin(baseModel, {
   timestamps: true,
 });
 
+schema.pre('init', function(next, chal) {
```
Do we need this for the group model as well?

Alys replied:
Yes, I'm planning to do much the same thing for guilds.

TheHollidayInn commented:
Looks good to me!

Alys committed:
finish implementating summary for challenges, add some support for guilds

Alys commented:
A screenshot of the Challenge creation form is below. The new summary feature is working for challenge creation. Things still to do:

- [copied to testing doc] Fix guild/party creation/edit flow so that correct information about group type and privacy is passed through to the edit form (currently I think there's bugs with that which pre-dated this PR). I suggest that fixing that is out of scope for this PR.
- [ ] Get summary fully working for public guilds and hidden on Tavern, private guilds, parties (requires first point above to be finished).
- [ ] Get categories fully working for public guilds and hidden on Tavern, private guilds, parties (requires first point above to be finished).
- [ ] Adjust existing tests that mention guild and challenge summaries to ensure they use it appropriately (I hacked the summaries in quickly to get the tests to pass).
- [copied to https://github.com/HabitRPG/habitica/issues/9245] Add tests for: 
   - [ ] creating a challenge through the API with NO summary provided - summary should equal challenge.name (necessary to ensure backwards compatibility for existing third-party tools)
   - [ ] creating a public guild through the API with NO summary provided - summary should equal group.name (necessary to ensure backwards compatibility for existing third-party tools)
   - [ ] creating a challenge with a summary provided - summary should be correct
   - [ ] creating a public guild with a summary provided - summary should be correct
   - [ ] other tests? Probably not but will think about it later.
- [x] "Public Challenges" text needs to be localised. How do you do that inside <script> ? EDIT: Oh duh. `this.$t('someString)`
- [copied to testing doc] Get challenge edit form working - it doesn't display summary or description. I suggest not doing that in this PR so we can get creation of challenges out faster for mods to test.
- [copied to testing doc] Get categories working in guild edit form - the current categories aren't displayed. 
- [copied to testing doc] Adjust API to check that the summary is not too long (the mongoose schema and vue file already do that) . I suggest not doing that in this PR since it's not urgent for website testing.
- [copied to testing doc] Improve error handling for user's mistakes when creating challenges (see @TODO comments in code and second screenshot below). I suggest not improving that in this PR since it's not as essential/urgent as other work.
- [ ] I'd also like to rename challengeModal.vue to challengeFormModal.vue for consistency with groupFormModal.vue but I'll do that in its own PR to make reviews easier.


![image](https://user-images.githubusercontent.com/1495809/29446399-5a2cf640-842f-11e7-98f0-602960809ea6.png)

![image](https://user-images.githubusercontent.com/1495809/29446451-9b246930-842f-11e7-97ec-91b88bd9406d.png)

Alys committed:
make small improvements to challenges code

Alys committed:
fix lint errors

Alys committed:
add more code to support summaries for guilds (still more needed)

Alys committed:
fix existing tests by adding summary field

Alys commented:
Assuming the tests pass, this is probably at a good stage for merging, if you don't have any complaints about how the code is written. (Complaints are welcome! I'm very new to vue.)

We still need more tests related to the summary fields which I will do, but I think I'll be too busy with personal events for the next day or so, and it might be best to merge this before conflicts occur. Other bugs in our testing document will require code that I've changed here to be modified in other ways, so conflicts would be likely.

There's also a fair bit more work to be done for guild creation/editing and a little more for challenges, but they're either things that are not urgent or that don't directly relate to the summary field and so are out of scope for this PR. I've updated the list of todos in my previous comment, which I guess we could move into the testing document, or at least link to from there. I think the best thing to do next is for someone (me? not me?) to look at the flow of data from the guild/party screens to the creation/edit form - I'm not sure we're passing through enough information about whether it's a public guild or private guild or party to be able to correctly control which fields and labels are shown, but I'm still not familiar enough with vue to be sure of what's happening. See the first todo in my previous post and the @TODO comments in this PR's code.

You're welcome to give me pointers about how to improve this, or push commits to my branch if you prefer.

Alys committed:
make existing tests pass

Alys committed:
Merge remote-tracking branch 'upstream/develop' into summary-for-guilds-challenges

negue commented on website/client/components/challenges/challengeModal.vue:
```diff
@@ -228,11 +234,11 @@ export default {
 
     this.groups.push({
-      name: 'Public',
+      name: 'Public Challenges',
```
could you extract this string ?

Alys replied:
Ah, thanks, yes, I should have noticed that! :) 

How do I do translatable text within a <script> section in a vue file? I've tried a quick search for other examples but can't find any and I have to get ready for work soon so I can't keep searching. :(

negue commented on website/client/components/challenges/challengeModal.vue:
```diff
@@ -285,26 +295,36 @@ export default {
-      this.workingChallenge.timestamp = new Date().getTime();
+      // @TODO: improve error handling, add it to updateChallenge, make errors translatable. Suggestion: `<% fieldName %> is required` where possible, where `fieldName` is inserted as the translatable string that's used for the field header.
+      let errors = '';
+      if (!this.workingChallenge.name) errors += 'Name is required\n';
```
could you extract this strings too ?

Alys replied:
No, this whole error handling section needs to be improved, as mentioned in that TODO comment so it would be best to create the translatable strings at the same time that the final error handling code is added (the strings currently here now aren't likely to stay). It's the same problem that exists in the group creation vue file, so ideally the same method of error handling would be added for both.

Alys committed:
WIP make "Public Challenges" text translatable

Alys committed:
Merge branch 'develop' of https://github.com/HabitRPG/habitrpg into summary-for-guilds-challenges

Alys committed:
change "leader" locale key to "guildOrPartyLeader" to make searches for it easier

Alys committed:
remove v-once from h2 headings

Alys commented:
My commits seem to have wiped out the review from @TheHollidayInn so from the email I got:

```
In website/client/components/challenges/challengeDetail.vue:
> @@ -56,6 +56,8 @@
       div(v-if='isLeader')
         button.btn.btn-danger(v-once, @click='closeChallenge()') {{$t('endChallenge')}}
     .description-section
+      h2(v-once) {{$t('challengeSummary')}}
Le't remove v-once for this. That was my mistake on the fields below. This prevents us from updating the detail view after using the edit modal
```
Done.


```
In website/client/components/groups/groupFormModal.vue:

> @@ -161,6 +168,8 @@ import toggleSwitch from 'client/components/ui/toggleSwitch';
 import gemIcon from 'assets/svg/gem.svg';
 import informationIcon from 'assets/svg/information.svg';
 
+import { MAX_SUMMARY_SIZE_FOR_GUILDS } from '../../../common/script/constants';
Maybe we should rename the constant to summary size? Is it the same for challenges and groups?
```
No, it's different. :) 250 for one, 500 for the other.


I've also fixed merge conflicts again, so I'm going to start testing it all again locally now, but once that's done and the travis build passes, can this be merged? If there's other non-critical changes I'm happy to do them, but the merge conflicts keep happening when other guild/challenge changes are made and I'm worried I'll eventually break the other changes if I don't resolve a merge conflict correctly. I'll post again here when I've finished my local tests.

Alys committed:
remove failed attempt to localise text in <script>

Alys committed:
add quick-and-dirty error checking for guild not having categories

Alys commented:
local testing finished, travis tests pass

Alys committed:
make "Public Challenges" text translatable

Alys committed:
rename final ...PlaceHolder strings to ...Placeholder (lower-case "h") for consistency with existing Placeholder strings

Alys commented:
I've fixed the problem with the "Public Challenges" text not being translatable.

The tests are passing. This is ready to merge if there's no significant problems with the code (minor problems I could fix in a new PR to avoid more merge conflicts with this one).
